### PR TITLE
[k8s]Create an IRuntimeInfo for Kubernetes.

### PR DESCRIPTION
**Pull request by darobs**
Create a IRuntimeInfoProvider implementation for Kubernetes pods.
Add some KubeUtils for the Runtime to use.
Add tests.

### CODE REVIEW

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+                return SanitizeDNSDomain(keySegments[0]) + "/" + SanitizeNameValue(keySegments[1]);
+            }
+            else
+            {
+                return SanitizeNameValue(key);
+            }
+        }
+
+        // Alphanumeric, '-' and '.' up to 253 characters.
+        public static string SanitizeK8sValue(string name)
```
[the remark this answers is not in the document]
Comment (darobs) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs]: Probably. I think we need to revisit this concept of just reducing the name to something acceptable to Kubernetes and think a little bit more about error handling.  I was less worried about it on this PR, but I can address it now.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+                imageName = status.Image;
+            }
+            if (status?.State?.Running != null)
+            {
+                if (status.State.Running.StartedAt.HasValue)
+                {
+                    return (0, Option.Some(status.State.Running.StartedAt.Value), Option.None<DateTime>(), imageName);
+                }
+            }
+            else if (status?.State?.Terminated != null)
```
[the remark this answers is not in the document]
Comment (darobs) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs]: Done

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+                    }
+                }
+                return (ModuleStatus.Unknown, "Unknown");
+            }).GetOrElse(() => (ModuleStatus.Unknown, "Unknown"));
+        }
+
+        static Option<V1ContainerStatus> GetContainerByName(string name, V1Pod pod)
+        {
+            string containerName = KubeUtils.SanitizeDNSValue(name);
+            if (pod.Status?.ContainerStatuses != null)
```
[the remark this answers is not in the document]
Comment (darobs) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs]: Looks better, thanks.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+                    Events.NullNodeInfoResponse(firstNode?.Metadata?.Name ?? "UNKNOWN");
+                } 
+            }
+            return new SystemInfo(osType, arch, version);
+        }
+
+        async Task WatchPodEventsAsync(WatchEventType type, V1Pod item)
+        {
+            // if the pod doesn't have the module label set then we are not interested in it
+            if (item.Metadata.Labels.ContainsKey(Constants.k8sEdgeModuleLabel) == false)
```
[the remark this answers is not in the document]
Comment (darobs) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs]: Agreed, fixed.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+            this.client = Preconditions.CheckNotNull(client, nameof(client));
+
+            this.podWatch = Option.None<Watcher<V1Pod>>();
+            this.moduleRuntimeInfos = new Dictionary<string, ModuleRuntimeInfo>();
+            this.moduleLock = new AsyncLock();
+        }
+
+        private async Task ListPodComplete(Task<HttpOperationResponse<V1PodList>> podListRespTask)
+        {
+            if (podListRespTask != null)
```
[the remark this answers is not in the document]
Comment (darobs) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs]: According to my design, there should only be one request operating at a time. but let me look at it more carefully.

File: Microsoft.Azure.Devices.Edge.sln [hunk 6]
```diff
@@ -177,13 +177,15 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Azure.Devices.Edg
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Azure.Devices.Edge.Test.TempSensor", "test\Microsoft.Azure.Devices.Edge.Test.TempSensor\Microsoft.Azure.Devices.Edge.Test.TempSensor.csproj", "{FCF94FF4-A608-4CC1-8D1D-215B6F88F203}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Microsoft.Azure.Devices.Edge.Agent.Kubernetes", "edge-agent\src\Microsoft.Azure.Devices.Edge.Agent.Kubernetes\Microsoft.Azure.Devices.Edge.Agent.Kubernetes.csproj", "{B63C506E-5CB5-4DE8-8275-F7BA8354799E}"
```
Comment (avranju) [inline Microsoft.Azure.Devices.Edge.sln]: Wonder why VS arbitrarily changes these GUIDS!

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/AgentEventIds.cs [hunk 7]
```diff
@@ -28,9 +28,10 @@ public struct AgentEventIds
         public const int KubernetesPlanner = EventIdStart + 2200;
         public const int KubernetesCommand = EventIdStart + 2300;
         public const int KubernetesOperator = EventIdStart + 2400;
-        public const int LogsUploadRequestHandler = EventIdStart + 2500;
-        public const int ModuleClientProvider = EventIdStart + 2600;
-        public const int LogsRequestHandler = EventIdStart + 2700;
+        public const int KubernetesReporter = EventIdStart + 2500;
+        public const int LogsUploadRequestHandler = EventIdStart + 2600;
+        public const int ModuleClientProvider = EventIdStart + 2700;
+        public const int LogsRequestHandler = EventIdStart + 2800;
```
Comment (avranju) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/AgentEventIds.cs]: Should we move the kubernetes constants into the kubernetes assembly instead of core?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+        const string AllowedCharsGeneric = "-.";
+        const int MaxK8SValueLength = 253;
+        const int MaxDnsNameLength = 63;
+        const int MaxLabelValueLength = 63;
+
+        static bool IsAlpha(char ch) => Alphabet.IndexOf(ch) != -1;
+
+        static bool IsAlphaNumeric(char ch) => IsAlpha(ch) || Char.IsDigit(ch);
+
+        public static readonly string K8sNamespace = Constants.k8sNamespaceBaseName;
```
Comment (avranju) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs]: Why do we need this string? Couldn't we just use `Constants.k8sNamespaceBaseName`?

Comment (avranju) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs]: Ah, it is re-assigned on line 35.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+            else
+            {
+                return SanitizeNameValue(key);
+            }
+        }
+
+        // Alphanumeric, '-' and '.' up to 253 characters.
+        public static string SanitizeK8sValue(string name)
+        {
+            name = name.ToLower();
```
Comment (avranju) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs]: Maybe we should do the `null` check we just did for the pub preview build.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+
+            var output = new StringBuilder();
+            for (int i = 0, len = 0; i < name.Length && len < MaxK8SValueLength; i++)
```
Comment (avranju) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs]: Do we need to track length separately in `len`? Wouldn't `output.Length` give us the length?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+            {
+                if (IsAlphaNumeric(name[i]) ||
+                    (len < MaxK8SValueLength-1 && AllowedCharsGeneric.IndexOf(name[i]) != -1))
```
Comment (avranju) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs]: Is the `len < MaxK8SValueLength-1` check necessary? As in, doesn't the loop terminator condition have the `len < MaxK8SValueLength` check? Why did we need the `-1`?

Also, nit: binary operators should have one space on each side.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/AgentEventIds.cs (hunk 7 above)
Comment (darobs) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/AgentEventIds.cs]: Yes we can.  I just need to make sure the EventIds are numerically unique.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs (hunk 11 above)
Comment (darobs) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs]: This is a result of the tests - There is a corner case if you have a a name with > 63 characters, but the 63rd character isn't valid to end a k8s name value, then I needed to throw out this character until we found an alphanumeric to end with.

This should be changed with @dmolokanov's suggestion that we throw if the name is too long, rather than truncate.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs (hunk 8 above)
Comment (darobs) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs]: Actually, I'm pulling this out, because the namespace will need to be passed in to anything that needs it.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+    {
+        const string Alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
+        const string AllowedCharsLabelValues = "-._";
+        const string AllowedCharsDns = "-";
+        const string AllowedCharsGeneric = "-.";
+        const int MaxK8SValueLength = 253;
+        const int MaxDnsNameLength = 63;
+        const int MaxLabelValueLength = 63;
+
+        static bool IsAlpha(char ch) => Alphabet.IndexOf(ch) != -1;
```
Comment (dmolokanov) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs]: What do you think about using hash set as a collection for alphabet? It will give us constant time of checking symbols against alphabet as opposite to linear scan in array?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+            if (output.Length > MaxK8SValueLength)
+            {
+                throw new InvalidKubernetesNameException($"Value '{name}' as sanitized exceeded maximum length {MaxK8SValueLength}");
+            }
+
+            return output.ToString();
+        }
+
+        // DNS label (as per RFC 1035)
+        public static string SanitizeDNSValue(string name, int maxLength = MaxDnsNameLength)
```
Comment (dmolokanov) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs]: Do we want to allow the caller of this method to pass `maxLength` in the method? Now, this method is somewhat generic and not very specific to k8s, right?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs (hunk 12 above)
Comment (darobs) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs]: Good idea!

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs (hunk 13 above)
Comment (darobs) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs]: This is an artifact from a previous iteration.  I should remove this.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+        static bool IsAlphaNumeric(char ch) => AlphaNumericHashSet.Contains(ch);
+
+        static KubeUtils()
+        {
+            AlphaHashSet = new HashSet<char>();
+            foreach (char c in Alphabet)
+            {
+                AlphaHashSet.Add(c);
+            }
+            AlphaNumericHashSet = new HashSet<char>(AlphaHashSet);
```
Comment (dmolokanov) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs]: I think instead of static ctor you can just initialize fields in-place like this:
```
static readonly HashSet<char> AlphaNumericHashSet = new HashSet<char>((Alphabet+Numeric).ToCharArray());
```
IMO it makes code more declarative and clear

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+
+            char[] annotationSplit = { '/' };
+            string[] keySegments = key.Split(annotationSplit, 2);
+            if (keySegments.Count() == 2)
+            {
+                return SanitizeDNSDomain(keySegments[0]) + "/" + SanitizeNameValue(keySegments[1]);
+            }
+            else
+            {
+                return SanitizeNameValue(key);
```
Comment (dmolokanov) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs]: On line 68 you throw an exception if a key is empty, but during sanitization you can end up with empty string. Is it ok for annotation key?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+                {
+                    output.Append(name[i]);
+                }
+            }
+            if (output.Length > MaxDnsNameLength)
+            {
+                throw new InvalidKubernetesNameException($"DNS name '{name}' exceeded maximum length of {MaxDnsNameLength}");
+            }
+
+            return output.ToString();
```
Comment (dmolokanov) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs]: Is it ok to have empty string after sanitization? I see the same pattern in each public function. Could you double check each of them, please?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs (hunk 14 above)
Comment (darobs) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs]: Nice.  Will do.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs [hunk 17]
```diff
@@ -0,0 +1,324 @@
...
+                            onClosed: () =>
+                            {
+                                Events.PodWatchClosed();
+
+                                // get rid of the current pod watch object since we got closed
+                                this.podWatch.ForEach(watch => watch.Dispose());
+                                this.podWatch = Option.None<Watcher<V1Pod>>();
+
+                                // kick off a new watch
+                                this.client.ListNamespacedPodWithHttpMessagesAsync(deviceNamespace, watch: true).ContinueWith(this.ListPodComplete);
```
Comment (kkdawkins) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs]: I really don't like the use of continuewith. I think it makes the code less readable and harder to debug. Why can't this be:
 
await this.client.ListNamespacedPodWithHttpMessagesAsync(deviceNamespace, watch: true);
this.ListPodComplete

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs [hunk 18]
```diff
@@ -0,0 +1,324 @@
...
+        static Option<V1ContainerStatus> GetContainerByName(string name, V1Pod pod)
+        {
+            string containerName = KubeUtils.SanitizeDNSValue(name);
+            return pod.Status?.ContainerStatuses
+                .Where(status => string.Equals(status.Name, containerName, StringComparison.OrdinalIgnoreCase))
+                .Select(status => Option.Some(status))
+                .FirstOrDefault() ?? Option.None<V1ContainerStatus>();
+        }
+
+        static (int, Option<DateTime>, Option<DateTime>, string image) GetTerminatedRuntimedata(V1ContainerStateTerminated term, string imageName)
```
Comment (kkdawkins) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs]: Can part of our work be moving from these (sometimes) massive tuples to C# objects? Like this could be an object called RuntimeData.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs [hunk 19]
```diff
@@ -0,0 +1,324 @@
...
+                                // kick off a new watch
+                                this.client.ListNamespacedPodWithHttpMessagesAsync(deviceNamespace, watch: true).ContinueWith(this.ListPodComplete);
+                            },
+                            onError: Events.ExceptionInPodWatch
+                        ));
+                }
+                else
+                {
+                    Events.NullListResponse("ListNamespacedPodWithHttpMessagesAsync", "http response");
+                    throw new NullReferenceException("Null response from ListNamespacedPodWithHttpMessagesAsync");
```
Comment (kkdawkins) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs]: Is this a nullref exception or is this a invalidresponse exception?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs (hunk 18 above)
Comment (kkdawkins) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs]: An advantage of C# objects is that new people to the codebase can quickly determine what the return values of a function are used for w/o needing to go back to the function to see where they came from :)

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs (hunk 16 above)
Comment (darobs) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs]: This may need to be tweaked as we need, but if it's throwing on null at the top, then it should also check for empty strings at the end.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs (hunk 17 above)
Comment (darobs) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs]: `await this.client.ListNamespacedPodWithHttpMessagesAsync()` blocks until a pod exists in the namespace, and we don't want to block the calling function.  If there's a classier way to do this, I would love to know it.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs [hunk 20]
```diff
@@ -0,0 +1,324 @@
...
+                }).GetOrElse(() => (ModuleStatus.Unknown, "Unknown"));
+        }
+
+        static Option<V1ContainerStatus> GetContainerByName(string name, V1Pod pod)
+        {
+            string containerName = KubeUtils.SanitizeDNSValue(name);
+            return pod.Status?.ContainerStatuses
+                .Where(status => string.Equals(status.Name, containerName, StringComparison.OrdinalIgnoreCase))
+                .Select(status => Option.Some(status))
+                .FirstOrDefault() ?? Option.None<V1ContainerStatus>();
```
Comment (kkdawkins) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs]: If it's default, will that be null? I ask because if it's an empty List then it will not trigger the Option.None

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs (hunk 19 above)
Comment (darobs) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs]: Added a new exception

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs (hunk 18 above)
Comment (darobs) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs]: I'm OK with leaving these as tuples, but I did make the change to see how it looks, and... it's OK.  There is an advantage in giving each field a name.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs (hunk 20 above)
Comment (darobs) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs]: The FirstOrDefault should return default(Option<V1ContainerStatus>), which is Option.None<V1ContainerStatus>.

**Comment by kkdawkins** [comment]
:shipit:
